### PR TITLE
Allow to customize foreign types in `#[derive(Associations)]`

### PR DESCRIPTION
**Pull request by weiznich**
This commit enables to usage of wrapper types for foreign keys in
`#[derive(Associations)]`
Basically those wrapper types must fulfil the following condition:
`&'a FieldType: Into<Option<&'a <ParentStruct as Identifiable>::Id>>`

### CODE REVIEW

File: diesel_derives/src/associations.rs [hunk 1]
```diff
@@ -35,26 +35,48 @@ fn derive_belongs_to(
...
     let struct_name = &model.name;
     let foreign_key_access = foreign_key_field.name.access();
     let foreign_key_ty = inner_of_option_ty(&foreign_key_field.ty);
     let table_name = model.table_name();
 
+    let mut generics = generics.clone();
+
+    // TODO: Remove this specialcasing as soon as we bump our minimal supported
+    // rust version to >= 1.30.0
+    // Then we could use the `&Option<T> -> `Option<&T>` from implementation in
```
Comment (sgrif) [inline diesel_derives/src/associations.rs]: Can this comment more explicitly say "Rust 1.30 adds `impl<'a, T> From<&'a Option<T>> for Option<&'a T>`"?

File: diesel_derives/src/associations.rs [hunk 2]
```diff
@@ -35,26 +35,48 @@ fn derive_belongs_to(
...
     };
 
+    generics.params.push(parse_quote!(__FK));
+    {
+        let where_clause = generics.where_clause.get_or_insert(parse_quote!(where));
+        where_clause
+            .predicates
+            .push(parse_quote!(__FK: std::hash::Hash + std::cmp::Eq));
+        where_clause.predicates.push(
+            parse_quote!(for<'__a> &'__a #foreign_key_ty: std::convert::Into<::std::option::Option<&'__a __FK>>),
```
Comment (sgrif) [inline diesel_derives/src/associations.rs]: This is much uglier than I had hoped XD (not that we can avoid it)

File: diesel_derives/src/associations.rs [hunk 3]
```diff
@@ -35,26 +35,48 @@ fn derive_belongs_to(
...
+        );
+    }
+
+    let (impl_generics, _, where_clause) = generics.split_for_impl();
+
     Ok(quote! {
         impl #impl_generics diesel::associations::BelongsTo<#parent_struct>
             for #struct_name #ty_generics
-        #where_clause
+            #where_clause
```
Comment (sgrif) [inline diesel_derives/src/associations.rs]: This seems unrelated?

**Comment by sgrif** [comment]
Needs a rebase or rebuild for green CI but looks good otherwise

**Comment by weiznich** [comment]
[That ci](https://travis-ci.org/diesel-rs/diesel/jobs/415328233#L544) error looks quite strange. I've played around a bit with the corresponding code:

```rust
fn _check(comment :Comment, posts: Post) {
    _test(&comment, &[posts])
}


pub fn _test<'a, T, P>(_t: &'a T, _p: &'a [P])
where &'a P: diesel::associations::Identifiable,
      T: HasTable + BelongsTo<P>,
      T::ForeignKeyColumn: ExpressionMethods,
      Vec<_Id<&'a P>>: AsInExpression<<T::ForeignKeyColumn as Expression>::SqlType>,
      <T as HasTable>::Table: FilterDsl<EqAny<T::ForeignKeyColumn, Vec<_Id<&'a P>>>>,
      T::ForeignKeyColumn: ExpressionMethods
{
    _test2(_t, _p)
}

pub fn _test2<'a, C, P>(_c: &'a C, _p: &'a [P]) where C: BelongingToDsl<&'a Vec<P>>
{}
```

does compile but 

```rust
fn _check(comment :Comment, posts: Post) {
    _test2(&comment, &[posts])
}


pub fn _test2<'a, C, P>(_c: &'a C, _p: &'a [P]) where C: BelongingToDsl<&'a Vec<P>>
{}
```

does not.

@diesel-rs/contributors any ideas?

File: diesel_derives/src/associations.rs [hunk 4]
```diff
@@ -35,26 +35,47 @@ fn derive_belongs_to(
...
     let table_name = model.table_name();
 
+    let mut generics = generics.clone();
+
+    // TODO: Remove this specialcasing as soon as we bump our minimal supported
+    // rust version to >= 1.30.0 because this version will add
+    // `impl<'a, T> From<&'a Option<T>> for Option<&'a T>` to the std-lib
     let foreign_key_expr = if is_option_ty(&foreign_key_field.ty) {
-        quote!(self#foreign_key_access.as_ref())
+        quote!(self#foreign_key_access.as_ref().and_then(std::convert::Into::into))
```
Comment (sgrif) [inline diesel_derives/src/associations.rs]: I don't think we want this change. It commits us to supporting `Option<&'a T> -> Option<&'a U>` for any type where `&'a T: Into<Option<&'a U>>`, which isn't enabled by your PR to Rust

File: diesel_derives/src/associations.rs [hunk 5]
```diff
@@ -35,26 +35,47 @@ fn derive_belongs_to(
...
 
     let foreign_key_field = model.find_column(&foreign_key)?;
     let struct_name = &model.name;
     let foreign_key_access = foreign_key_field.name.access();
     let foreign_key_ty = inner_of_option_ty(&foreign_key_field.ty);
     let table_name = model.table_name();
 
+    let mut generics = generics.clone();
+
+    // TODO: Remove this specialcasing as soon as we bump our minimal supported
```
Comment (sgrif) [inline diesel_derives/src/associations.rs]: typo here, "special casing" is two words (surprised clippy didn't catch this)

File: examples/postgres/getting_started_step_3/src/lib.rs [hunk 6]
```diff
@@ -15,16 +15,14 @@ pub fn establish_connection() -> PgConnection {
...
 }
 
 pub fn create_post(conn: &PgConnection, title: &str, body: &str) -> Post {
     use schema::posts;
 
-    let new_post = NewPost {
-        title: title,
-        body: body,
-    };
+    let new_post = NewPost { title, body };
```
Comment (sgrif) [inline examples/postgres/getting_started_step_3/src/lib.rs]: I've just been allowing this elsewhere to avoid the git churn, but I guess it's not too bad if it's only in two places.
